File: src/main/java/dev/alexengrig/leetcode/_1_two_sum/MapSolution.java
package dev.alexengrig.leetcode._1_two_sum;

import java.util.HashMap;
import java.util.Map;

class MapSolution implements Task {
    @Override
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            Integer value = map.get(nums[i]);
            if (value != null) {
                return new int[]{value, i};
            }
            map.put(target - nums[i], i);
        }
        throw new IllegalArgumentException(); // unreachable
    }
}

File: src/main/python/1_two_sum/map_solution.py
from typing import List

from task import Task


class MapSolution(Task):
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        seen = {}
        for i, num in enumerate(nums):
            if num in seen:
                return [seen[num], i]
            seen[target - num] = i
        raise ValueError("No solution")
